Core/src/Surfaces/DiscSurface.cpp

```diff
@@ -373,3 +376,120 @@ double Acts::DiscSurface::pathCorrection(const GeometryContext& gctx,
   /// we can ignore the global position here
   return 1. / std::abs(normal(gctx).dot(direction));
 }
+
+std::shared_ptr<Acts::DiscSurface> Acts::DiscSurface::mergedWith(
+    const GeometryContext& gctx, const DiscSurface& other,
+    BinningValue direction, const Logger& logger) const {
+  using namespace Acts::UnitLiterals;
+
+  ACTS_DEBUG("Merging disc surfaces in " << binningValueNames()[direction]
+                                         << " direction")
+
+  Transform3 otherLocal = transform(gctx).inverse() * other.transform(gctx);
+
+  constexpr auto tolerance = s_onSurfaceTolerance;
+
+  // surface cannot have any relative rotation
+  if (!otherLocal.linear().isApprox(RotationMatrix3::Identity())) {
+    ACTS_ERROR("DiscSurface::merge: surfaces have relative rotation");
+    throw std::invalid_argument(
+        "DiscSurface::merge: surfaces have relative rotation");
+  }
+
+  Vector3 translation = otherLocal.translation();
+
+  if (!translation.isApprox(Vector3::Zero())) {
+    ACTS_ERROR(
+        "DiscSurface::merge: surfaces have relative translation in x/y/z");
+    throw std::invalid_argument(
+        "DiscSurface::merge: surfaces have relative translation in x/y/z");
+  }
```

Comment:
in the `CylinderSurface` implementation we use the surface tolerance. Should we align that to use one of both?